tokio/src/time/driver/sleep.rs

```diff
@@ -191,11 +212,43 @@ pin_project! {
 }
 
 impl Sleep {
-    pub(crate) fn new_timeout(deadline: Instant) -> Sleep {
-        let handle = Handle::current();
-        let entry = TimerEntry::new(&handle, deadline);
+    cfg_not_trace! {
+        pub(crate) fn new_timeout(deadline: Instant) -> Sleep {
+            let handle = Handle::current();
+            let entry = TimerEntry::new(&handle, deadline);
+            Sleep { inner: Inner {deadline}, entry }
+        }
+    }
+
+    cfg_trace! {
+        pub(crate) fn new_timeout(deadline: Instant) -> Sleep {
+            let handle = Handle::current();
+            let entry = TimerEntry::new(&handle, deadline);
+
+            let time_source =  handle.time_source().clone();
+            let duration = time_source.deadline_to_tick(deadline) - time_source.now();
+            let resource_span = tracing::trace_span!(
+                "runtime.resource",
+                concrete_type = "Sleep",
+                kind = "timer");
+
+            let async_op_span = tracing::trace_span!(
+                "runtime.resource.async_op",
+                source = "Sleep::new_timeout"
```

Comment:
@hawkw  I think that makes sense. My initial thought was that the async op span represents the life of the future that represents the async operation. So really from the moment of instantiating the future to the time it is dropped. This is also why there is this little comment in the proto code: https://github.com/tokio-rs/console/blob/main/proto/async_ops.proto#L52. Namely, until it is polled, we do not really know which task or which resource this op is associated with. I think what you are describing makes sense, especially the suggestion around the "source" field. One reason I went with creating the span as part of the future instantiation is that I would assume that ideally we would like to be able to observe a situation where a bunch of futures are created by are never polled? WDYT?

Reply:
Hmm. I think the issue is that a `Sleep` is _not_ necessarily a single `Future`. It might also be constructed internally as part of an `Interval`, and I think it's important that every time the interval is `tick`ed is considered a separate async op. 

I could go either way on when the async op span should _begin_, but I think distinguishing between awaiting a `Sleep` as a future, and `tick`ing an `Interval` is important.

Reply:
You could have it reset the span every time `Sleep::reset` is called?

Reply:
> You could have it reset the span every time `Sleep::reset` is called?

Yeah, that plus maybe passing in a string constant for the "source" field to distinguish between spans created via `time::sleep` and `time::interval` would probably be sufficient!

Reply:
@hawkw All of this makes me wonder whether the source field is useful at all. Passing string constants around seems a bit brittle. I wonder if there is a more elegant way to represent the fact that the sleep resource is being used by the interval? This is not the only instance of the pattern.  For example `RwLock` internally uses a `Semaphore`. Should we just get rid of the "source" field an introduce a parent-child relationship between resources where a the main resource span can be a child of another resource span.  Or event rely on the stack of entered spans to determine that and instead of relying on a field rely on the backtrace of entered spans? This way we will be able to associate the fact that this async op (and all others) that are happening on the Sleep future are being triggered by the parent `Interval` resource. Does that make sense? Does it sound better? The fact that we need to reset the async op span when resetting the timer is still true.

Reply:
Hmm, I think that in the eventual future, establishing the relationships between higher and lower-level async operations using a span hierarchy definitely does make sense. However, it's also additional work we'd have to do in the subscriber. I think that _having_ the `source` field is probably fine though, especially since adding code to the console subscriber to infer these relationships based on inspecting the span context would require additional changes.

In general, I think it's fine to add as many fields to things as we like, and not worry *too* much about changing them around later...as long as they're _just_ displayed in the "fields" section, and don't have other special meanings to the console subscriber. If we _do_ add special meanings to fields beyond just displaying them as arbitrary k/v pairs, then we need to be a little more cautious about removing/changing them, but I think it's fine to keep the `source` field and continue iterating on it in further branches, especially if it lets us move forward with this and come back to it later...